WebCrypto X25519 algorithm docs (#35441)
mdn/content: files/en-us/web/api/subtlecrypto/derivebits/index.md

});
```

### X25519

> In this live example Alice and Bob each generate an X25519 key pair.

Reviewer: ```suggestion
In this example Alice and Bob each generate an X25519 key pair.
```

(just a suggestion. I doubt most readers know what "live" means here, and it ought to be obvious they can interact with it.)

We then use Alice's private key and Bob's public key to derive a secret, and compare that with the secret generated using Bobs's private key and Alices's public key to show that they are shared/identical.

#### HTML

> The HTML is just a button that an be pressed to generate all the keys and the shared key.

Reviewer: This doesn't seem right. Two buttons, one derives a shared secret with the two keys pairs, the other generates new key pairs.

Author: Grrr. Changed it half way through.

```suggestion
The HTML is defines two buttons.
The "Change keys" button can be pressed to generate new key pairs for Alice and Bob.
The "Derive bits" button can be pressed to derive a shared secret with the current set of key pairs.
```

  return await window.crypto.subtle.deriveBits(
    {
      name: "X25519",
      public: publicKey,
    },
    privateKey,
    128,
  );
}
```

The code below adds a function to generate new keys for Alice and Bob.
This is done the first time the JavaScript is loaded, and repeated whenever the "Change keys" button is pressed (this allows us to see the effect of changing the keys on the shared secret).

```js
let alicesKeyPair;
let bobsKeyPair;

async function changeKeys() {
  try {
    alicesKeyPair = await window.crypto.subtle.generateKey(
      {
        name: "X25519",
      },
      false,
      ["deriveBits"],
    );

    bobsKeyPair = await window.crypto.subtle.generateKey(
      {
        name: "X25519",
      },
      false,
      ["deriveBits"],
    );
  } catch (e) {
    log(e);
  }
}

changeKeys();

const changeKeysButton = document.querySelector("#buttonChangeKeys");

changeKeysButton.addEventListener("click", async () => {
  // Generate 2 ECDH key pairs: one for Alice and one for Bob
  // In more normal usage, they would generate their key pairs
  // separately and exchange public keys securely
  changeKeys();

>   log("Keys changed");

Reviewer: The async is a bit messed up here. If you try this in a browser that doesn't support X25519 (like Safari) you get:

```
Keys changed
NotSupportedError: The operation is not supported.
``` 

...because `changeKeys()` returns a Promise, and we log "Keys changed" before the promise fulfills.

I think it might work better if you take the try...catch out of `changeKeys()`, an put it here instead, like:

```js
try {
  await changeKeys();
  log(Keys changed");
}
catch(e) {
  log(e);
}
```

I didn't test this though.

Author: Good catch - this always reports key change even if on fail. 

I fixed this by moving the log() into the success-path of the existing try-catch.
That way it is after an `await` and only gets run when the promise has fulfilled.

Chrome doesn't support this algorithms and displays:

> NotSupportedError: Failed to execute 'generateKey' on 'SubtleCrypto': Algorithm: Unrecognized name

});
```

> The code below adds a handler function that is invoked every time the "Derive-bits" button is pressed.

Reviewer: ```suggestion
The code below adds a handler function that is invoked every time the "Derive bits" button is pressed.
```

The handler generates the shared secrets for Alice and Bob using the `deriveSharedSecret()` method defined above, and logs them for easy comparison.

```js
const deriveBitsButton = document.querySelector("#buttonDeriveKeys");

deriveBitsButton.addEventListener("click", async () => {

>   // Generate 2 ECDH key pairs: one for Alice and one for Bob

Author: ```suggestion
  // Generate 2 X25519 key pairs: one for Alice and one for Bob
```

  // In more normal usage, they would generate their key pairs
  // separately and exchange public keys securely

  // Alice then generates a secret using her private key and Bob's public key.
  // Bob could generate the same secret using his private key and Alice's public key.

  const sharedSecretAlice = await deriveSharedSecret(
    alicesKeyPair.privateKey,
    bobsKeyPair.publicKey,
  );

  let buffer = new Uint8Array(sharedSecretAlice, 0, 10);
  log(`${buffer}…[${sharedSecretAlice.byteLength} bytes total] (Alice secret)`);

  const sharedSecretBob = await deriveSharedSecret(
    bobsKeyPair.privateKey,
    alicesKeyPair.publicKey,
  );

  buffer = new Uint8Array(sharedSecretBob, 0, 10);
  log(`${buffer}…[${sharedSecretAlice.byteLength} bytes total] (Bob secret)`);
});
```

#### Result

> Press the "Derive-bits" button to generate and log a shared secret from Bob and Alices keys.

Reviewer: ```suggestion
Press the "Derive bits" button to generate and log a shared secret from Bob and Alice's keys.
```

> Press the "Change keys" button to change the keys used by both parties.

Reviewer: ```suggestion
Press the "Change keys" button to change the X25519 keys used by both parties.
```

Maybe? To be clear about which keys we are changing.

  logElement.scrollTop = logElement.scrollHeight;
}
```

The function to generate a shared secret using the X25519 algorithm is shown below.
This takes a private key from one party and the public key from another.

```js
async function deriveSharedSecret(privateKey, publicKey) {
  return await window.crypto.subtle.deriveBits(
    {
      name: "X25519",
      public: publicKey,
    },
    privateKey,
    128,
  );
}
```

The code below adds a function to generate new keys for Alice and Bob.
This is done the first time the JavaScript is loaded, and repeated whenever the "Change keys" button is pressed (this allows us to see the effect of changing the keys on the shared secret).

```js
let alicesKeyPair;
let bobsKeyPair;

async function changeKeys() {
  try {
    alicesKeyPair = await window.crypto.subtle.generateKey(
      {
        name: "X25519",
      },
      false,
      ["deriveBits"],
    );

    bobsKeyPair = await window.crypto.subtle.generateKey(
      {
        name: "X25519",
      },
      false,
      ["deriveBits"],
    );

    log("Keys changed");
  } catch (e) {
    log(e);
  }
}

changeKeys();

const changeKeysButton = document.querySelector("#buttonChangeKeys");

> changeKeysButton.addEventListener("click", async () => {
>   // Generate 2 X25519 key pairs: one for Alice and one for Bob
>   // In more normal usage, they would generate their key pairs
>   // separately and exchange public keys securely
>   changeKeys();
> });

Reviewer: I think at this point you can just do this:

```suggestion
// Generate 2 X25519 key pairs: one for Alice and one for Bob
// In more normal usage, they would generate their key pairs
// separately and exchange public keys securely
changeKeysButton.addEventListener("click", changeKeys);
```

can't you?

Author: You're right!
